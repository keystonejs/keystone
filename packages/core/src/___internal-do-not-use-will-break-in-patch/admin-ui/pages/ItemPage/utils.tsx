import { useMemo } from 'react'
import {
  getRootGraphQLFieldsFromFieldController,
} from '../../../../admin-ui/utils'
import isDeepEqual from 'fast-deep-equal'
import type {
  FieldMeta,
  JSONValue
} from '../../../../types'

export function deserializeItemToValue (
  fields: Record<string, FieldMeta>,
  item: Record<string, unknown | null>
) {
  const result: Record<string, unknown | null> = {}
  for (const fieldKey in fields) {
    const field = fields[fieldKey]
    const itemForField: Record<string, unknown> = {}
    for (const graphqlField of getRootGraphQLFieldsFromFieldController(field.controller)) {
      itemForField[graphqlField] = item?.[graphqlField] ?? null
    }
    result[fieldKey] = field.controller.deserialize(itemForField)
  }
  return result
}

function serializeValueToItem (
  fields: Record<string, FieldMeta>,
  value: Record<string, unknown>
) {
  const result: Record<string, Record<string, JSONValue>> = {}
  for (const fieldKey in fields) {
    const fieldValue = value[fieldKey]
    result[fieldKey] = fields[fieldKey].controller.serialize(fieldValue)
  }
  return result
}

export function useChangedFieldsAndDataForUpdate (
  fields: Record<string, FieldMeta>,
  item: Record<string, unknown>,
  value: Record<string, unknown>,
) {
  const serializedItem = useMemo(() => serializeValueToItem(fields, value), [fields, value])
  const serializedValuesFromItem = useMemo(() => {
    const value2 = deserializeItemToValue(fields, item)
    return serializeValueToItem(fields, value2)
  }, [fields, item])

  return useMemo(() => {
    const changedFields = new Set<string>()
    for (const fieldKey in serializedItem) {
      const isEqual = isDeepEqual(
        serializedItem[fieldKey],
        serializedValuesFromItem[fieldKey]
      )
      if (!isEqual) {
        changedFields.add(fieldKey)
      }
    }

    const dataForUpdate: Record<string, any> = {}
    for (const fieldKey of changedFields) {
      Object.assign(dataForUpdate, serializedItem[fieldKey])
    }

    Object.keys(serializedItem)
      .filter(fieldKey => fields[fieldKey].graphql.isNonNull?.includes('update'))
      .filter(fieldKey => !changedFields.has(fieldKey))
      .forEach(fieldKey => {
        Object.assign(dataForUpdate, serializedItem[fieldKey])
      })

    return {
      hasChangedFields: changedFields.size > 0,
      dataForUpdate
    }
  }, [serializedItem, serializedValuesFromItem, fields])
}
