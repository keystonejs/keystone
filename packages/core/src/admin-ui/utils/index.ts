export * from './Fields'
export * from './getRootGraphQLFieldsFromFieldController'
export * from './useInvalidFields'

export type DeepNullable<T> =
  | null
  | (T extends Array<infer Item>
      ? Array<DeepNullable<Item>>
      : { [Key in keyof T]: DeepNullable<T[Key]> })
