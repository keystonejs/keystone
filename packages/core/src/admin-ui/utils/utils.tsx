import { useMemo } from 'react'
import isDeepEqual from 'fast-deep-equal'
import type {
  FieldMeta,
} from '../../../types'
import weakMemoize from '@emotion/weak-memoize'
import {
  type FragmentDefinitionNode,
  type SelectionSetNode,
  parse,
} from 'graphql'
import type { FieldController } from '../../types'

function extractRootFields (selectedFields: Set<string>, selectionSet: SelectionSetNode) {
  selectionSet.selections.forEach(selection => {
    if (selection.kind === 'Field') {
      selectedFields.add(selection.alias ? selection.alias.value : selection.name.value)
    }
    if (selection.kind === 'InlineFragment') {
      extractRootFields(selectedFields, selection.selectionSet)
    }
    // FragmentSpread will never happen for the use cases of getRootFieldsFromSelection
  })
}

export const getRootGraphQLFieldsFromFieldController = weakMemoize(
  (controller: FieldController<any, any>) => {
    const ast = parse(`fragment X on Y {
  id
  ${controller.graphqlSelection}
  }`)
    const selectedFields = new Set<string>()
    const fragmentNode = ast.definitions[0] as FragmentDefinitionNode
    extractRootFields(selectedFields, fragmentNode.selectionSet)
    return [...selectedFields]
  }
)

export function useInvalidFields (
  fields: Record<string, FieldMeta>,
  item: Record<string, unknown>
): ReadonlySet<string> {
  return useMemo(() => {
    const invalidFields = new Set<string>()

    for (const fieldKey in item) {
      const validateFn = fields[fieldKey]?.controller?.validate
      if (!validateFn) continue

      const fieldValue = item[fieldKey]
      const valid = validateFn(fieldValue)
      if (valid) continue

      invalidFields.add(fieldKey)
    }
    return invalidFields
  }, [fields, item])
}

export function makeDefaultValueState (
  fields: Record<string, FieldMeta>,
) {
  const result: Record<string, unknown> = {}
  for (const fieldKey in fields) {
    result[fieldKey] = fields[fieldKey].controller.defaultValue
  }
  return result
}

export function deserializeItemToValue (
  fields: Record<string, FieldMeta>,
  item: Record<string, unknown | null>
) {
  const result: Record<string, unknown | null> = {}
  for (const fieldKey in fields) {
    const field = fields[fieldKey]
    const itemForField: Record<string, unknown> = {}
    for (const graphqlField of getRootGraphQLFieldsFromFieldController(field.controller)) {
      itemForField[graphqlField] = item?.[graphqlField] ?? null
    }
    result[fieldKey] = field.controller.deserialize(itemForField)
  }
  return result
}

export function serializeValueToOperationItem (
  operation: 'create' | 'update',
  fields: Record<string, FieldMeta>,
  value: Record<string, unknown>,
  valueReference?: Record<string, unknown>
) {
  const result: Record<string, unknown> = {}
  valueReference ??= makeDefaultValueState(fields)

  for (const fieldKey in fields) {
    const field = fields[fieldKey]

    const fieldValue = value[fieldKey]
    const fieldValueSerialized = field.controller.serialize(fieldValue)

    const fieldValueReference = valueReference[fieldKey]

    const isAlwaysRequired = field.graphql.isNonNull.includes(operation)
    if (!isAlwaysRequired) {
      if (isDeepEqual(fieldValueSerialized, field.controller.serialize(fieldValueReference))) continue
    }

    Object.assign(result, fieldValueSerialized)
  }

  return result
}

export function useHasChanges (
  operation: 'create' | 'update',
  fields: Record<string, FieldMeta>,
  value: Record<string, unknown>,
  valueReference?: Record<string, unknown>
) {
  return useMemo(() => {
    const alwaysRequiredCount = Object.values(fields)
      .filter(f => f.graphql.isNonNull.includes(operation))
      .length

    const itemForUpdate = serializeValueToOperationItem(operation, fields, value, valueReference)

    // add any fields that are always required
    return (Object.keys(itemForUpdate).length - alwaysRequiredCount) > 0
  }, [fields, value, valueReference])
}
