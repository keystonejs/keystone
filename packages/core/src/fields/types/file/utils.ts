import { AssetMode } from '../../../types';

const FILEREGEX = /^(local|cloud):file:([^\\\/:\n]+)/;

export const getFileRef = (mode: AssetMode, name: string) => `${mode}:file:${name}`;
export const parseFileRef = (ref: string) => {
  const match = ref.match(FILEREGEX);
  if (match) {
    const [, mode, filename] = match;
    return {
      mode: mode as AssetMode,
      filename: filename as string,
    };
  }
  return undefined;
};
