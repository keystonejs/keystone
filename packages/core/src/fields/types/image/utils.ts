export const SUPPORTED_IMAGE_EXTENSIONS = ['jpg', 'png', 'webp', 'gif'];
