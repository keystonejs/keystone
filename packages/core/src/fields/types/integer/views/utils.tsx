import { ChangeEvent, FocusEvent, useState } from 'react';

type ParsedValueBase = undefined | symbol | boolean | object | number | null | bigint;

type Config<ParsedValue extends ParsedValueBase> = {
  parse: (value: string) => ParsedValue | string;
  format: (value: ParsedValue) => string;
};

export function useFormattedInput<ParsedValue extends ParsedValueBase>(
  config: Config<ParsedValue>,
  {
    value,
    onChange,
    onBlur,
    onFocus,
  }: {
    value: string | ParsedValue;
    onChange: (value: string | ParsedValue) => void;
    onFocus?: (event: FocusEvent<HTMLInputElement>) => void;
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  }
) {
  // typeof value === 'string' implies the unparsed form
  // typeof value !== 'string' implies the parsed form
  if (typeof value === 'string' && typeof config.parse(value) !== 'string') {
    throw new Error(
      `Valid values must be passed in as a parsed value, not a raw value. The value you passed was \`${JSON.stringify(
        value
      )}\`, you should pass \`${JSON.stringify(config.parse(value))}\` instead`
    );
  }
  let [internalValueState, setInternalValueState] = useState(() =>
    typeof value === 'string' ? value : config.format(value)
  );
  const [isFocused, setIsFocused] = useState(false);
  if (typeof value === 'string' && value !== internalValueState) {
    setInternalValueState(value);
  }
  // If the value is not a string, we know it's in the parsed form
  if (typeof value !== 'string') {
    const formatted = config.format(value);
    // When the input is blurred, we want to show always show the formatted
    // version so if we're not focussed and the formatted version is different
    // to the current version, we need to update it.
    if (!isFocused && formatted !== internalValueState) {
      setInternalValueState(formatted);
    }

    const parsedInternal = config.parse(internalValueState);

    // We updating the internal value here because the
    // external value has changed.
    if (typeof parsedInternal !== 'string' && config.format(parsedInternal) !== formatted) {
      setInternalValueState(formatted);
    }
  }

  return {
    value: internalValueState,
    onChange(event: ChangeEvent<HTMLInputElement>) {
      const value = event.target.value;
      const parsed = config.parse(value);
      onChange(parsed);
      setInternalValueState(value);
    },
    onFocus(event: FocusEvent<HTMLInputElement>) {
      onFocus?.(event);
      setIsFocused(true);
    },
    onBlur(event: FocusEvent<HTMLInputElement>) {
      onBlur?.(event);
      setIsFocused(false);
      // this isn't strictly necessary since we already do this in render
      // this just saves another rerender after setIsFocused(false)
      if (typeof value !== 'string') {
        setInternalValueState(config.format(value));
      }
    },
  };
}
