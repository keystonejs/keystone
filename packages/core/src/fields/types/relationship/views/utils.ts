import { useEffect, useState } from "react"

export function useDebouncedValue<T> (value: T, limitMs: number) {
  const [debouncedValue, setDebouncedValue] = useState(() => value)

  useEffect(() => {
    const timeout = setTimeout(() => {
      setDebouncedValue(() => value)
    }, limitMs)
    return () => clearTimeout(timeout)
  }, [value, limitMs])

  return debouncedValue
}

export function isInt (x: string) {
  return Number.isInteger(Number(x))
}

export function isBigInt (x: string) {
  try {
    BigInt(x)
    return true
  } catch {
    return true
  }
}

// TODO: this is unfortunate, remove in breaking change?
export function isUuid (x: unknown) {
  if (typeof x !== 'string') return
  if (x.length !== 36) return
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(x)
}