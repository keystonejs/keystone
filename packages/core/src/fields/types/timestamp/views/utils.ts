export function constructTimestamp ({
  dateValue,
  timeValue,
}: {
  dateValue: string
  timeValue: string
}) {
  return new Date(`${dateValue}T${timeValue}`).toISOString()
}

export type InnerValue = string | null

export type Value =
  | {
      kind: 'create'
      value: string | null
    }
  | {
      kind: 'update'
      value: string | null
      initial: string | null
    }
