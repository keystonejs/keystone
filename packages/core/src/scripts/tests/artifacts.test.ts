import fs from 'fs/promises';
import { ExitError } from '../utils';
import {
  basicKeystoneConfig,
  getFiles,
  recordConsole,
  runCommand,
  schemas,
  symlinkKeystoneDeps,
  testdir,
} from './utils';

describe.each(['postinstall', 'build', 'prisma migrate status'])('%s', command => {
  test('logs an error and exits with 1 when the schemas do not exist and the terminal is non-interactive', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    });
    const recording = recordConsole();
    await expect(runCommand(tmp, command)).rejects.toEqual(new ExitError(1));
    expect(recording()).toMatchInlineSnapshot(`
      "Your Prisma and GraphQL schemas are not up to date
      Please run keystone postinstall --fix to update your Prisma and GraphQL schemas"
    `);
  });
});

const schemasMatch = ['schema.prisma', 'schema.graphql'];

// a lot of these cases are also the same for prisma and build commands but we don't include them here
// because when they're slow and then run the same code as the postinstall command
// (and in the case of the build command we need to spawn a child process which would make each case take a _very_ long time)
describe('postinstall', () => {
  test('prompts when in an interactive terminal to update the schema', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    });
    const recording = recordConsole({
      'Would you like to update your Prisma and GraphQL schemas?': true,
    });
    await runCommand(tmp, 'postinstall');
    const files = await getFiles(tmp, schemasMatch);
    expect(files).toEqual(await getFiles(`${__dirname}/fixtures/basic-project`, schemasMatch));
    expect(recording()).toMatchInlineSnapshot(`
      "Your Prisma and GraphQL schemas are not up to date
      Prompt: Would you like to update your Prisma and GraphQL schemas? true
      ✨ GraphQL and Prisma schemas are up to date"
    `);
  });
  test('updates the schemas without prompting when --fix is passed', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    });
    const recording = recordConsole();
    await runCommand(tmp, 'postinstall --fix');
    const files = await getFiles(tmp, schemasMatch);
    expect(files).toEqual(await getFiles(`${__dirname}/fixtures/basic-project`, schemasMatch));
    expect(recording()).toMatchInlineSnapshot(`"✨ Generated GraphQL and Prisma schemas"`);
  });
  test("does not prompt, error or modify the schemas if they're already up to date", async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': basicKeystoneConfig,
    });
    const recording = recordConsole();
    await runCommand(tmp, 'postinstall');
    const files = await getFiles(tmp, schemasMatch);
    expect(files).toEqual(await getFiles(`${__dirname}/fixtures/basic-project`, schemasMatch));
    expect(recording()).toMatchInlineSnapshot(`"✨ GraphQL and Prisma schemas are up to date"`);
  });
  test('writes the correct node_modules files', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': basicKeystoneConfig,
    });
    const recording = recordConsole();
    await runCommand(tmp, 'postinstall');
    expect(await getFiles(tmp, ['node_modules/.keystone/**/*'])).toMatchSnapshot();
    expect(recording()).toMatchInlineSnapshot(`"✨ GraphQL and Prisma schemas are up to date"`);
  });
  test('writes the api files when the generateNodeAPI experimental flag is on', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': await fs.readFile(`${__dirname}/fixtures/generate-node-api.ts`, 'utf8'),
    });
    const recording = recordConsole();
    await runCommand(tmp, 'postinstall');
    expect(await getFiles(tmp, ['node_modules/.keystone/api.{d.ts,js}'])).toMatchInlineSnapshot(`
      ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/api.d.ts ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
      import { KeystoneListsAPI } from '@keystone-6/core/types';
      import { Context } from './types';

      export const query: Context['query'];
      ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/api.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
      import keystoneConfig from '../../keystone';
      import { PrismaClient } from '.prisma/client';
      import { createQueryAPI } from '@keystone-6/core/___internal-do-not-use-will-break-in-patch/node-api';
      import path from 'path';

          path.join(__dirname, "../../../app.db");
          path.join(process.cwd(), "app.db");
          

      export const query = createQueryAPI(keystoneConfig, PrismaClient);

    `);
    expect(recording()).toMatchInlineSnapshot(`"✨ GraphQL and Prisma schemas are up to date"`);
  });
  test('writes the next graphql api route files when the generateNextGraphqlAPI experimental flag is on', async () => {
    const tmp = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': await fs.readFile(
        `${__dirname}/fixtures/generate-next-graphql-api.ts`,
        'utf8'
      ),
    });
    const recording = recordConsole();
    await runCommand(tmp, 'postinstall');
    expect(await getFiles(tmp, ['node_modules/.keystone/next/graphql-api.{d.ts,js}']))
      .toMatchInlineSnapshot(`
      ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/next/graphql-api.d.ts ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
      export const config: any;
      export default config;

      ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/next/graphql-api.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
      import keystoneConfig from '../../../keystone';
      import { PrismaClient } from '.prisma/client';
      import { nextGraphQLAPIRoute } from '@keystone-6/core/___internal-do-not-use-will-break-in-patch/next-graphql';
      import path from 'path';

          path.join(__dirname, "../../../app.db");
          path.join(process.cwd(), "app.db");
          

      export const config = {
        api: {
          bodyParser: false,
        },
      };

      export default nextGraphQLAPIRoute(keystoneConfig, PrismaClient);

    `);
    expect(recording()).toMatchInlineSnapshot(`"✨ GraphQL and Prisma schemas are up to date"`);
  });
});
