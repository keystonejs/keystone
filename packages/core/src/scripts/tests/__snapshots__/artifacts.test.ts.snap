// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postinstall writes the correct node_modules files 1`] = `
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/types.d.ts ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
type Scalars = {
  readonly ID: string;
  readonly Boolean: boolean;
  readonly String: string;
  readonly Int: number;
  readonly Float: number;
  readonly JSON: import("@keystone-6/core/types").JSONValue;
  readonly Decimal: import("@keystone-6/core/types").Decimal | string;
};

export type TodoWhereUniqueInput = {
  readonly id?: Scalars["ID"] | null;
};

export type TodoWhereInput = {
  readonly AND?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null;
  readonly OR?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null;
  readonly NOT?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null;
  readonly id?: IDFilter | null;
  readonly title?: StringFilter | null;
};

export type IDFilter = {
  readonly equals?: Scalars["ID"] | null;
  readonly in?: ReadonlyArray<Scalars["ID"]> | Scalars["ID"] | null;
  readonly notIn?: ReadonlyArray<Scalars["ID"]> | Scalars["ID"] | null;
  readonly lt?: Scalars["ID"] | null;
  readonly lte?: Scalars["ID"] | null;
  readonly gt?: Scalars["ID"] | null;
  readonly gte?: Scalars["ID"] | null;
  readonly not?: IDFilter | null;
};

export type StringFilter = {
  readonly equals?: Scalars["String"] | null;
  readonly in?: ReadonlyArray<Scalars["String"]> | Scalars["String"] | null;
  readonly notIn?: ReadonlyArray<Scalars["String"]> | Scalars["String"] | null;
  readonly lt?: Scalars["String"] | null;
  readonly lte?: Scalars["String"] | null;
  readonly gt?: Scalars["String"] | null;
  readonly gte?: Scalars["String"] | null;
  readonly contains?: Scalars["String"] | null;
  readonly startsWith?: Scalars["String"] | null;
  readonly endsWith?: Scalars["String"] | null;
  readonly not?: NestedStringFilter | null;
};

export type NestedStringFilter = {
  readonly equals?: Scalars["String"] | null;
  readonly in?: ReadonlyArray<Scalars["String"]> | Scalars["String"] | null;
  readonly notIn?: ReadonlyArray<Scalars["String"]> | Scalars["String"] | null;
  readonly lt?: Scalars["String"] | null;
  readonly lte?: Scalars["String"] | null;
  readonly gt?: Scalars["String"] | null;
  readonly gte?: Scalars["String"] | null;
  readonly contains?: Scalars["String"] | null;
  readonly startsWith?: Scalars["String"] | null;
  readonly endsWith?: Scalars["String"] | null;
  readonly not?: NestedStringFilter | null;
};

export type TodoOrderByInput = {
  readonly id?: OrderDirection | null;
  readonly title?: OrderDirection | null;
};

export type OrderDirection =
  | "asc"
  | "desc";

export type TodoUpdateInput = {
  readonly title?: Scalars["String"] | null;
};

export type TodoUpdateArgs = {
  readonly where: TodoWhereUniqueInput;
  readonly data: TodoUpdateInput;
};

export type TodoCreateInput = {
  readonly title?: Scalars["String"] | null;
};

export type KeystoneAdminUIFieldMetaCreateViewFieldMode =
  | "edit"
  | "hidden";

export type KeystoneAdminUIFieldMetaListViewFieldMode =
  | "read"
  | "hidden";

export type KeystoneAdminUIFieldMetaItemViewFieldMode =
  | "edit"
  | "read"
  | "hidden";

export type QueryMode =
  | "default"
  | "insensitive";

export type KeystoneAdminUISortDirection =
  | "ASC"
  | "DESC";

type ResolvedTodoCreateInput = {
  id?: import('.prisma/client').Prisma.TodoCreateInput["id"];
  title?: import('.prisma/client').Prisma.TodoCreateInput["title"];
};

type ResolvedTodoUpdateInput = {
  id?: import('.prisma/client').Prisma.TodoUpdateInput["id"];
  title?: import('.prisma/client').Prisma.TodoUpdateInput["title"];
};


export declare namespace Models {
  export type Todo = import('@keystone-6/core').ListConfig<Models.Todo.TypeInfo, any>;
  namespace Todo {
    export type Item = import('.prisma/client').Todo;
    export type TypeInfo = {
      key: "Todo";
      fields: "id" | "title"
      item: Item;
      inputs: {
        where: TodoWhereInput;
        uniqueWhere: TodoWhereUniqueInput;
        create: TodoCreateInput;
        update: TodoUpdateInput;
        orderBy: TodoOrderByInput;
      };
      prisma: {
        create: ResolvedTodoCreateInput
        update: ResolvedTodoUpdateInput
      };
      all: __TypeInfo;
    };
  }
}
export type Context = import('@keystone-6/core/types').KeystoneContext<TypeInfo>;

export type TypeInfo = {
  lists: {
  readonly Todo: Models.Todo.TypeInfo;
  };
  prisma: import('.prisma/client').PrismaClient;
};

type __TypeInfo = TypeInfo;

export type Models = {
  [Key in keyof TypeInfo['lists']]?: import('@keystone-6/core').ListConfig<TypeInfo['lists'][Key], any>
} & Record<string, import('@keystone-6/core').ListConfig<any, any>>;

export {}

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/types.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

`;
